enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+
+- A Machine should be paired 1:1 with an instance on the cloud provider
+  - If the instance is preempted/terminated, the cloud actuator should not replace it
+  - If the instance is preempted/terminated, the cloud provider should not replace it
+
+- The actuator is responsible for creation of the instance only and should not attempt to remediate problems
+
+- The actuator should not attempt to verify that an instance can be created before attempting to create the instance
+  - If the cloud provider does not have capacity, the Machine Health Checker can (given required MHC) remove the Machine after a period.
+    MachineSet will ensure the correct number of Machines are created.
+
+- There’s a [2 hours time frame](https://github.com/openshift/cluster-machine-approver/blob/master/csr_check.go#L38) in OpenShift for a machine to become a node.
+  If the Spot request is not satisfied during this time frame, it is assumed that the instance will never become a node.
+  This machine would be remediated by a given MHC.
+
+
+### Implementation Details
+
+#### Cloud Provider Implementation Specifics
+
+##### AWS
+
+###### Launching instances
+
+To launch an instance as a Spot instance on AWS, a [SpotMarketOptions](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#SpotMarketOptions)
+needs to be added to the `RunInstancesInput`. Within this there are 3 options that matter:
+
+- InstanceInterruptionBehaviour (default: terminate): This must be set to `terminate` otherwise the SpotInstanceType cannot be `one-time`
+
+- SpotInstanceType (default: one-time): This must be set to `one-time` to ensure that each Machine only creates on EC2 instance and that the spot request is
+
+- MaxPrice (default: On-Demand price): This can be **optionally** set to a string representation of the hourly maximum spot price.
+  If not set, the option will default to the On-Demand price of the EC2 instance type requested
+
+The only option from this that needs exposing to the user from this is the `MaxPrice`, this option should be in an optional struct, if the struct is not nil,
+then spot instances should be used, if the MaxPrice is set, this should be used instead of the default On-Demand price.
+
+```
+type SpotMarketOptions struct {
+  MaxPrice *string `json:”maxPrice,omitempty”`
+}
+
+type AWSMachineProviderConfig struct {
+  ...
+
+  SpotMarketOptions *SpotMarketOptions `json:”spotMarketOptions,omitempty”`
+}
+```
+
+###### Termination Notices
+Termination notices on AWS are provided via the EC2 metadata service up to 2 minutes before an instance is due to be preempted.
+A DaemonSet should be deployed on the cluster which runs Pods on nodes labelled `machine.openshift.io/terminate-on-preemption: true`.
+Users of Spot instances could optionally add this label to their Machine’s spec to opt in to this feature.
+
+The DaemonSet itself would poll the EC2 metadata service for a termination notice.
```

Comment:
If we give the daemonset privilege to delete the machine, we have very high privileged pod running on a worker.

Maybe set a health-check that fails on the daemonset when we get the 200 OK from metadata, and have a controller watch pods in a particular namespace.

Reply:
I think i like the require MHC if you want to use spot instances and termination triggers the MHC to fail early approach more than deleting the machine object..

Reply:
In that line the daemonset could set the node conditions the MHC is configured to watch.

Reply:
I think there are several ways that this could be achieved, depending on how we want to break down the privileges that the pods have and how much complexity we want.
- Daemonset that has privileges to delete machines (as currently proposed)
  - Potentially dangerous as highly privileged, bad actor could delete all machines
- Daemonset that has privileges to set a condition on any node, for MHC to notice
  - If the MHC sees this condition, should it terminate immediately? If not, can we afford a delay?
  - Is this any different from above? Bad actor could apply condition to all nodes and by proxy, delete all Machines
- Daemonset that uses Node's credentials to add condition to Node
  - MHC would have to immediately act on this condition
  - Using Node account limits bad actors capabilities to just the node that the pod is on, safer?
  - Is reusing Node account ok? Grants more privileges than necessary, not intended use, could break in future?
- Some (to be defined) controller creates a Pod for each node with it's own role and service account
  - Can limit privileges to only be able to delete the Machine for the Node the Pod is running on
  - Alternatively could only be able to update status of Node Pod is running on
  - More controllers, more complexity than other solutions

Reply:
I'm proposing a daemonset to schedule a pod on each spot node.  The daemonset simply polls the metadata service.  If it gets the shutdown notification, fail the readiness check for that pod.  Have a controller watching pods in whatever namespace the daemonset is in.  When a pod fails readiness (after an initial startup delay), that controller deletes the associated machine.

I think this might work, that you mentioned:

> Daemonset that has privileges to set a condition on any node, for MHC to notice

The pod is going to have access to instance metadata, host networking, so it might as well have an account that lets it set conditions on the node.  Is it possible to give it permissions to only set conditions on the local node?  Having it able to mark this condition on other nodes would be very bad.

MHC would need to respond immediately without delay.  This should be simple enough to configure.

Reply:
> I'm proposing a daemonset to schedule a pod on each spot node. The daemonset simply polls the metadata service. If it gets the shutdown notification, fail the readiness check for that pod. Have a controller watching pods in whatever namespace the daemonset is in. When a pod fails readiness (after an initial startup delay), that controller deletes the associated machine.

My only concern with this is adding additional delays that cannot really be afforded when you have a short notice period. If the health check period for the pod is say, 5s, and the pod itself polls the metadata every 5s, you could easily lose 10s of a 30s period on GCP or Azure.

> The pod is going to have access to instance metadata, host networking, so it might as well have an account that lets it set conditions on the node. Is it possible to give it permissions to only set conditions on the local node? Having it able to mark this condition on other nodes would be very bad.

The third and fourth options in my [comment](https://github.com/openshift/enhancements/pull/199/files#r376312656) above have permissions to restrict to only setting conditions on a single node, I think the fourth option is likely to be the most complex to implement, but has the best security aspect if we are giving any permissions to the pods doing the health checking that is.

> MHC would need to respond immediately without delay. This should be simple enough to configure.

Ack, agreed!

Reply:
GCP sends a signal, so presumably we'd catch that almost immediately.

I think we have a variety of options.  Most of them are fine with me, as long as we're not running a really privileged pod on each node (eg, a pod that can delete machines).  We'll probably figure out which method is best by implementing and iterating.

Reply:
I've added a [detailed explanation](ttps://github.com/openshift/enhancements/pull/199/files#diff-c0faaddbbe335546befe28b0664694f9R83-R136) of the way I think we should proceed with this. It's a little more complex than I had originally hoped but should be relatively safe in terms of not giving pods too much privilege.

I am wondering if it should be a separate proposal now though due to its complexity 🤔

Reply:
Let's try to keep this simple initially. I think current system design already give us all we need:
- Actuators must signal new machines/nodes as a spot instances, i.e label it with `machine.openshift.io/spot` (name to be decided).
- MAO deploys a new daemonSet with a nodeSelector that matches the well know spot label.
- The provider image/binary to run by the controller own by the new daemonSet is chosen by mao per cloud (as we already do for actuators. The source code can indeed live along with each actuator).
- This controller does its provider specific magic to realise imminent termination and signal the machine for deletion (or if there're concerns with that, it sets a node condition and delegates on the optional existence of a MHC to signal the deletion based on the conditions).